src/d3a/models/strategy/storage.py

```diff
@@ -127,61 +131,108 @@ def __init__(self, initial_soc: float = StorageSettings.MIN_ALLOWED_SOC,
             self.bid_update.update_interval = update_interval
             self.offer_update.update_interval = update_interval
 
     def area_reconfigure_event(self, **kwargs):
         self._area_reconfigure_prices(**kwargs)
         self.offer_update.update_on_activate()
         self.bid_update.update_on_activate()
-        self._validate_rates()
 
-    def _validate_rates(self):
-        for time_slot in generate_market_slot_list():
-            bid_rate_change = None if self.bid_update.fit_to_limit else \
-                self.bid_update.energy_rate_change_per_update[time_slot]
-            offer_rate_change = None if self.offer_update.fit_to_limit else \
-                self.offer_update.energy_rate_change_per_update[time_slot]
-            validate_storage_device(initial_selling_rate=self.offer_update.initial_rate[time_slot],
-                                    final_selling_rate=self.offer_update.final_rate[time_slot],
-                                    initial_buying_rate=self.bid_update.initial_rate[time_slot],
-                                    final_buying_rate=self.bid_update.final_rate[time_slot],
-                                    energy_rate_increase_per_update=bid_rate_change,
-                                    energy_rate_decrease_per_update=offer_rate_change,
-                                    fit_to_limit=self.bid_update.fit_to_limit,
-                                    update_interval=self.bid_update.update_interval)
+    @staticmethod
+    def _validate_rates(initial_selling_rate, final_selling_rate,
+                        initial_buying_rate, final_buying_rate,
+                        energy_rate_increase_per_update, energy_rate_decrease_per_update,
+                        bid_fit_to_limit, offer_fit_to_limit,
+                        init=False):
+        try:
+            for time_slot in generate_market_slot_list():
+                bid_rate_change = None if bid_fit_to_limit else \
+                    energy_rate_increase_per_update[time_slot]
+                offer_rate_change = None if offer_fit_to_limit else \
+                    energy_rate_decrease_per_update[time_slot]
+                validate_storage_device(initial_selling_rate=initial_selling_rate[time_slot],
+                                        final_selling_rate=final_selling_rate[time_slot],
+                                        initial_buying_rate=initial_buying_rate[time_slot],
+                                        final_buying_rate=final_buying_rate[time_slot],
+                                        energy_rate_increase_per_update=bid_rate_change,
+                                        energy_rate_decrease_per_update=offer_rate_change,
+                                        fit_to_limit=bid_fit_to_limit)
+            return True
+        except D3AException as e:
+            if init:
+                raise e
+            else:
+                log.error(str(e))
+                return False
```

Comment:
I do not like that much the mixture of boolean return value and exception raise logic. Why not just raising an exception and dealing with it in the calling code ? If I see correctly, you only need boolean return values in the area_reconfigure_event methods. Consequently you can catch the exceptions there and either leave the parameters unchanged in case an exception is raised, or change the full parameter list if no exception is raised. You can easily do that using a try/except/else block.

Reply:
True! You are right. The motivation here was to put all work into one function, bu ti agree with you and will change it accordingly.